autoload/ale/c.vim

```diff
@@ -89,3 +89,45 @@ function! ale#c#FindCompileCommands(buffer) abort
 
     return ''
 endfunction
+
+" Given a buffer number, find a compilation database (compile_commands.json)
+" and parse this into an argument string
+function! ale#c#FindCompileArgs(buffer) abort
+    let l:compile_args = {}
+
+    let l:compile_commands_path = ale#path#FindNearestFile(a:buffer, 'compile_commands.json')
+    if !empty(l:compile_commands_path)
+        let l:bufname = fnamemodify(bufname(a:buffer), ':p')
```

Comment:
The buffer name can be different from the filename for the buffer. You should use `expand('#' . a:buffer . ':p')` instead.